File: corporate-reporting/src/types/Milestone.ts
import { IEntity } from "./Entity";
import { IUser } from "./User";
import { IAttribute } from "./Attribute";
import { IKeyWorkArea } from "./KeyWorkArea";
import { IWorkStream } from "./WorkStream";
import { IReportingEntityWithDeliveryDates, ReportingEntityWithDeliveryDates } from "./ReportingEntity";
import { IPartnerOrganisation } from "./PartnerOrganisation";
import { IMilestoneUpdate } from "./MilestoneUpdate";

export interface IMilestone extends IReportingEntityWithDeliveryDates {
    MilestoneCode: string;
    StartDate: Date;
    MilestoneTypeID: number;
    Attributes: IAttribute[];
    LeadUserID: number;
    RagOptionID: number;
    WorkStreamID: number;
    KeyWorkAreaID: number;
    PartnerOrganisationID: number;
    LeadUser?: IUser;
    MilestoneType?: IEntity;
    KeyWorkArea?: IKeyWorkArea;
    WorkStream?: IWorkStream;
    PartnerOrganisation?: IPartnerOrganisation;
    Description: string;
    MilestoneUpdates?: IMilestoneUpdate[];
}

export class Milestone extends ReportingEntityWithDeliveryDates implements IMilestone {
    public MilestoneCode = '';
    public StartDate = null;
    public MilestoneTypeID = null;
    public Attributes = [];
    public LeadUserID = null;
    public RagOptionID = null;
    public WorkStreamID = null;
    public KeyWorkAreaID = null;
    public PartnerOrganisationID = null;
    public Description = '';
}

File: ORB.Core/Models/Milestone.cs
using System;
using System.Collections.Generic;

namespace ORB.Core.Models
{
    public class Milestone : ReportingSubEntity
    {
        public Milestone()
        {
            MilestoneUpdates = new HashSet<MilestoneUpdate>();
        }
        
        public string MilestoneCode { get; set; }
        public DateTime? BaselineDate { get; set; }
        public DateTime? ForecastDate { get; set; }
        public DateTime? ActualDate { get; set; }
        public int? MilestoneTypeID { get; set; }
        public int? RagOptionID { get; set; }
        public int? WorkStreamID { get; set; }
        public int? KeyWorkAreaID { get; set; }
        public int? PartnerOrganisationID { get; set; }
        public DateTime? StartDate { get; set; }

        public virtual KeyWorkArea KeyWorkArea { get; set; }
        public virtual MilestoneType MilestoneType { get; set; }
        public virtual PartnerOrganisation PartnerOrganisation { get; set; }
        public virtual RagOption RagOption { get; set; }
        public virtual WorkStream WorkStream { get; set; }
        public virtual ICollection<MilestoneUpdate> MilestoneUpdates { get; set; }
    }
}
